drivers/serial/uart_ra.c

```diff
@@ -0,0 +1,396 @@
+	uart_ra_write_8(dev, SCR, reg_val);
+
+	data->current_config = *cfg;
+
+	k_spin_unlock(&data->lock, key);
+
+	return 0;
+}
+
+#ifdef CONFIG_UART_USE_RUNTIME_CONFIGURE
+static int uart_ra_config_get(const struct device *dev,
+			      struct uart_config *cfg)
+{
+	struct uart_ra_data *data = dev->data;
+
+	*cfg = data->current_config;
+
+	return 0;
+}
+#endif /* CONFIG_UART_USE_RUNTIME_CONFIGURE */
+
+static int uart_ra_init(const struct device *dev)
+{
+	const struct uart_ra_cfg *config = dev->config;
+	struct uart_ra_data *data = dev->data;
+	int ret;
+
+	/* Configure dt provided device signals when available */
+	ret = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (!device_is_ready(config->clock_dev)) {
+		return -ENODEV;
+	}
+
+	ret = clock_control_on(config->clock_dev, config->clock_id);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = clock_control_get_rate(config->clock_dev, config->clock_id, &data->clk_rate);
+	if (ret < 0) {
+		return ret;
+	}
+
+	DEVICE_MMIO_MAP(dev, K_MEM_CACHE_NONE);
+
+	ret = uart_ra_configure(dev, &data->current_config);
+	if (ret != 0) {
+		return ret;
+	}
+
+	return 0;
+}
+
+static const struct uart_driver_api uart_ra_driver_api = {
+	.poll_in = uart_ra_poll_in,
+	.poll_out = uart_ra_poll_out,
+#ifdef CONFIG_UART_USE_RUNTIME_CONFIGURE
+	.configure = uart_ra_configure,
+	.config_get = uart_ra_config_get,
+#endif
+};
+
+/* Device Instantiation */
+#define UART_RCAR_INIT_CFG(n)                                                                      \
+	PINCTRL_DT_DEFINE(DT_INST_PARENT(n));                                                      \
+	static const struct uart_ra_cfg uart_ra_cfg_##n = {                                        \
+		.regs = DT_REG_ADDR(DT_INST_PARENT(n)),                                            \
+		.clock_dev = DEVICE_DT_GET(DT_CLOCKS_CTLR(DT_INST_PARENT(n))),                     \
+		.clock_id =                                                                        \
+			(clock_control_subsys_t)DT_CLOCKS_CELL_BY_IDX(DT_INST_PARENT(n), 0, id),   \
+		.pcfg = PINCTRL_DT_DEV_CONFIG_GET(DT_INST_PARENT(n)),                              \
+	}
+
+#define UART_RCAR_INIT(n)							\
+	static struct uart_ra_data uart_ra_data_##n = {				\
+		.current_config = {						\
+			.baudrate = DT_INST_PROP(n, current_speed),		\
+			.parity = UART_CFG_PARITY_NONE,				\
+			.stop_bits = UART_CFG_STOP_BITS_1,			\
+			.data_bits = UART_CFG_DATA_BITS_8,			\
+			.flow_ctrl = UART_CFG_FLOW_CTRL_NONE,			\
+		},								\
+	};									\
+										\
+	static const struct uart_ra_cfg uart_ra_cfg_##n;			\
+										\
+	DEVICE_DT_INST_DEFINE(n,						\
+			      uart_ra_init,					\
+			      NULL,						\
+			      &uart_ra_data_##n,				\
+			      &uart_ra_cfg_##n,					\
+			      PRE_KERNEL_1, CONFIG_SERIAL_INIT_PRIORITY,	\
+			      &uart_ra_driver_api);				\
+										\
+	UART_RCAR_INIT_CFG(n);
```

Comment:
maybe move this before the `DEVICE_DT_INST_DEFINE`? Then you can drop the forward declaration.